torch/csrc/api/include/torch/nn/modules/sequential.h

```diff
@@ -196,17 +245,30 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
   /// `Sequential(std::make_shared<Module>(3, 4))`.
   template <typename M, typename = torch::detail::enable_if_module_t<M>>
   void push_back(M&& module) {
-    // Need to get rid of any reference components for make_unique.
-    using Type = typename std::remove_reference<M>::type;
-    // Here we move (or copy) the module into a new shared_ptr.
-    push_back(std::make_shared<Type>(std::forward<M>(module)));
+    push_back(NamedSubmodule(std::to_string(modules_.size()), std::forward<M>(module)));
   }
 
   /// Unwraps the contained module of a `ModuleHolder` and adds it to the
   /// `Sequential`.
   template <typename M>
   void push_back(const ModuleHolder<M>& module_holder) {
-    push_back(module_holder.ptr());
+    push_back(NamedSubmodule(std::to_string(modules_.size()), module_holder));
+  }
+
+  /// Adds a new `NamedSubmodule` to the `Sequential` container.
+  /// This method allows creating the `NamedSubmodule` from braced-init-list, e.g.
+  /// `push_back({"linear", Linear(3, 4)})`
+  void push_back(NamedSubmodule named_module) {
```

Comment:
can you not provide push_backs for (string, module)?  NamedSubmodule being part of the public interface feels clunky, because what you get back from named_submodule isn't a NamedSubmodule at all!  It's just an OrderedDict of modules, which you can handle perfectly well.